firebase-appdistribution/test-app/src/main/java/com/googletest/firebase/appdistribution/testapp/ScreenshotDetectionFeedbackTrigger.kt

```diff
@@ -0,0 +1,226 @@
+package com.googletest.firebase.appdistribution.testapp
+
+import android.Manifest.permission.READ_EXTERNAL_STORAGE
+import android.Manifest.permission.READ_MEDIA_IMAGES
+import android.annotation.SuppressLint
+import android.app.Activity
+import android.app.AlertDialog
+import android.app.Application
+import android.content.pm.PackageManager.PERMISSION_GRANTED
+import android.database.ContentObserver
+import android.database.Cursor
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.os.Handler
+import android.os.HandlerThread
+import android.provider.MediaStore
+import android.util.Log
+import androidx.activity.result.ActivityResultCaller
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.core.content.ContextCompat
+import com.google.firebase.appdistribution.FirebaseAppDistribution
+import java.util.*
+
+class ScreenshotDetectionFeedbackTrigger
+private constructor(private val infoTextResourceId: Int, handler: Handler) :
+  ContentObserver(handler), Application.ActivityLifecycleCallbacks {
+
+  private val seenImages = HashSet<Uri>()
+  private var requestPermissionLauncher: ActivityResultLauncher<String>? = null
+  private var currentActivity: Activity? = null
+  private var currentUri: Uri? = null
+  private var isEnabled = false
+  private var hasRequestedPermission = false
+
+  override fun onChange(selfChange: Boolean, uri: Uri?) {
+    if (uri == null || !isExternalContent(uri) || seenImages.contains(uri)) {
+      return
+    }
+    currentActivity?.let { activity ->
```

Comment:
I think http://go/kotlin/practices/scope-functions#avoid-using-let-where-smart-casts-suffice applies here.

Maybe you can just add `|| currentActivity == null` to the `if` above and Kotlin's "smart cast" can figure out that `currentActivity` is not null below?

Reply:
Unfortunately I couldn't use smart cast here, because the `currentActivity` member could in theory have been set to null by some other thread in between the null check and the usage.

BUT I just realized another way to get around this is using functions that take a nullable argument. That argument becomes a local reference that I can then use with a smart cast.

We could talk about the possible issues from different threads updating and using currentActivity but I don't think that poses any major problems. There's a slim chance we try to use a reference to an activity that's already been hidden or destroyed. There won't be any NPEs or anything.

Reply:
Thanks for explaining!